### PR TITLE
fix(ai,settings): surface provider reasons in AI connection tests

### PR DESCRIPTION
**Pull request by theBGuy**
A failed **Test connection** in Settings → AI showed only the HTTP reason phrase ("Bad Request"), which reads like a malformed request even when the real cause was a rejected key. This surfaces the provider's own explanation, clears stale test results when the key changes, and stops suggesting Gemini models a free AI Studio key can't call.

### AI client error reporting

- Adds `providerErrorMessage()` in `src/lib/ai/client.ts`, which parses the unread `responseBody` on the SDK's `APICallError` and returns `error.message` from it — handling Google's array-wrapped payload and the bare object OpenAI-compatible servers return, and falling back to the generic `errorMessage(e)` for a non-JSON body.
- Routes all three failure paths through it: the `"error"` stream part, the streaming catch block, and `testConnection()`'s result message.

### Settings → AI panel

- Clears `testResult` after saving a key and after removing one in `src/features/settings/AiProviderSection.tsx`, so a result describing the old key no longer makes a fixed setup look broken (previously only a provider/model change cleared it).
- Trims the unsaved key before building the client in the test handler, matching what Save stores so a pasted key with stray whitespace doesn't fail the test and then work once saved.

### Gemini model suggestions

- Replaces `gemini-2.5-pro` with `gemini-3.5-flash-lite` in `GEMINI_MODELS` (`src/lib/ai/providers.ts`), making the suggested list Flash-tier only; the doc comment records that Pro answers a free AI Studio key with 429 `limit: 0`, while paid keys still reach Pro through the live catalog.

### Changelog

- Adds `changelog.d/fixed-ai-connection-test-feedback.md` and `changelog.d/changed-gemini-model-suggestions.md` covering the two user-facing changes.

### CODE REVIEW

**Comment by theBGuy** [comment]
## Context for reviewers

**1. What this is.** Four fixes to how AI provider failures surface, plus the Gemini suggestion list. All of it is fallout from #183: a live debugging session on `master` produced a false "master is broken" report, and the root cause turned out to be a **stale test-result banner**, not a regression. This PR fixes what made that misdiagnosis possible.

**2. `providerErrorMessage()` exists because the SDK hides the cause.** When a provider's error body doesn't match the SDK's expected schema, `APICallError.message` is set to the bare HTTP reason phrase and the real reason sits unread in `responseBody`. Measured against the live API: Google answers a rejected key with **400** and `[{"error":{"message":"Invalid Auth key."}}]` — array-wrapped, and **not** 401. That is exactly why the app said "Bad Request".

**3. The two unwraps in that helper are load-bearing, not defensive padding.**
- **Retries:** `ai@6.0.225` wraps exhausted retries in a `RetryError` (`ai/dist/index.mjs:2725`) whose body lives on `lastError` (`:531`); `maxRetries` defaults to 2 (`:2711`) and `shouldRetry` fires on `isRetryable` (`:2720`), which is true for 429/5xx. Without this unwrap the most common failure class of all — quota — still reports "Failed after 3 attempts".
- **In-stream `error` parts** arrive as the provider's already-parsed payload rather than an `Error`, so `String(e)` would render `[object Object]`.

**4. OpenAI is unaffected by design.** Its bodies match the SDK schema (`errorToMessage` puts the text on `.message`), so the helper is a no-op there — verified with a junk key against `api.openai.com`. This targets providers whose body shape the SDK can't parse.

**5. The stale-banner fix needs a generation guard, not just a clear.** Clearing on save/clear alone still loses a race: a test started against the *old* key can resolve after the clear and re-pin its verdict (and the mirror case shows a green **Connected** under *"(no key saved)"*). `testRun` invalidates in-flight runs. `setTesting(false)` is deliberately **unconditional** — the button is disabled while testing, so no newer run can own that flag, and gating it would strand the button disabled forever.

**6. Gemini suggestions are Flash-only because Pro is unusable on a free key.** Measured with a real free-tier AI Studio key: `gemini-3.6-flash` / `gemini-3.5-flash` / `gemini-3.5-flash-lite` → **200**; `gemini-2.5-pro` **and** `gemini-3.1-pro-preview` → **429 `limit: 0`** — tier-wide, not one stale id. The live catalog still lists Pro: it shows what a key can *see*, not what its tier can *call*. Existing saved models are untouched (the picker free-types, and `defaultModelForProvider` only fires on a provider switch).

**7. `"google/gemini-2.5-pro"` surviving in the diff is not an oversight** — that's the Vercel AI Gateway routing slug, a different path from a direct Google key.

### Disclosures

**8. No doc surface changed, deliberately.** `rg -i "test connection"` across `README.md`, the in-app guide and `site/src` returns zero hits, and no surface enumerates suggested model ids. Nothing stale and nothing to add — a recorded call, not a silent omission. Two changelog fragments cover the user-facing half.

**9. `review-tools.ts` (8 sites) and `cli-client.ts` (2) keep `errorMessage` on purpose.** Those are tool-execution and subprocess errors with no `responseBody` to read. The class boundary here is "AI-SDK HTTP provider errors", not a file boundary — all 6 sites inside that boundary are converted.

**10. Known gap, with a home.** The live **model-catalog** fetch swallows its failure and then shows *"Suggestions only — save an API key to load the live list"* even when a key **is** saved but rejected. Fixing it needs a failure reason threaded out of `useAvailableModels`, which is a bigger change than this PR's scope; it's on the backlog rather than left unnoticed.

**11. Local `biome ci ./src/` noise is environmental.** It flags 20 files here, none of them in this diff — CRLF from the branch checkout. CI runs the same command on an LF checkout and is green on this head.


---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

Sound change: the provider-error unwrap is grounded (I confirmed `RetryError.lastError` at `ai@6.0.225` `dist/index.mjs:531`, and that the openai chat path enqueues `{type:"error", error: value.error}` — the parsed payload, so the `message` fallback is the right shape), and the generation guard on the test result is correct — `disabled={testing}` at `AiProviderSection.tsx:866` really does make the unconditional `setTesting(false)` safe. Nothing blocking; one staleness gap remains in the same class the PR sets out to close.

### Correctness / edge cases

- **should-fix** — `AiProviderSection.tsx`, `allowHost` (line ~700) and the `AllowedHostsField` `onChange` (line ~1060): the draft allowlist is a real input to the test (`testConnection` passes `allowedHosts` → `createAiClient` → `guardedFetch(allowedOverride)`, `guarded-fetch.ts:22-40`), but neither writer calls `discardTestResult()`. Concrete case: provider `ollama`/`openai-compatible` with a custom host, Test connection → red banner "GitDesktop blocked a network request to …"; the user clicks the adjacent **Allow host** chip, which routes `HostAllowNote:45 → onAllowHost → allowHost` and only touches `aiAllowedHosts` (it never goes through `setAi`), so the red verdict stays pinned over a setup that is now allowed — exactly the "fixed setup looks broken" case the fragment promises. Mirror case: remove the host in `AllowedHostsField` after a green test and **Connected** survives. Fix: call `discardTestResult()` after the `form.setFieldValue("aiAllowedHosts", …)` in `allowHost`, and convert the line-1060 arrow to a block that does `form.setFieldValue("aiAllowedHosts", next); discardTestResult();`. Both are in scope of the function, so no imports change. Note the third writer, `McpServersSection.tsx:127`, edits the same list from a component with no access to this state — that hole stays open unless the result state is lifted; worth stating as a deliberate call rather than leaving it silent.

- **nit** — `AiProviderSection.tsx`, the `key` field (`keyForm.AppField name="key"`, ~line 809): a test run against a typed-but-unsaved key keeps its verdict while the user retypes the key, since only `onSubmit` discards. A `listeners={{ onChange: discardTestResult }}` on that field would close it; if you'd rather not (the fragment only promises clearing on save/remove), that's a reasonable scope line.

- **nit** — `client.ts:56`: the final fallback returns `errorMessage(e)`, the *wrapper*, not `errorMessage(unwrapped)`. Harmless today only because provider-utils builds the retry message as `Failed after N attempts. Last error: <msg>` (`@ai-sdk/provider-utils/dist/index.mjs:2672`), so the last error's text still reaches the UI behind that prefix — worth a conscious "we keep the attempt count" rather than an accident.

### Readability

- **nit** — `client.ts:22-31`, `providerErrorMessage` doc comment: "Two unwraps are load-bearing…" argues the diff rather than stating a constraint (CLAUDE.md: never argue the diff is correct), and the second path isn't an unwrap — it's the top-level `message` fallback for in-stream error parts. Trim to the two constraints: Google's array-wrapped body leaves the real reason in `responseBody` while `.message` is the HTTP reason phrase, and retries move that body onto `RetryError.lastError`.

### Docs

The recorded "no doc surface changed" call checks out against what I can see: `rg` over `README.md`, `src/features/help/content.ts` and `site/src` turns up Gemini only as a provider name (`README.md:817`, `content.ts:1783`, `capabilities.ts:641`) — no suggested-model ids and no "Test connection" copy to go stale. Both fragments match `changelog.d/README.md`'s format (leading `- `, 2-space continuation), and `CHANGELOG.md` is correctly left untouched.

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI security audit** · `opus` · automated

---

No exploitable security issues found in these changes: the provider error text is rendered as React text (escaped) or, on markdown paths, through DOMPurify-sanitized `Markdown`; the JSON parsing in `providerErrorMessage()` only reads properties (no merge/assign, so no prototype pollution), and the test-run generation guard and model-id list changes carry no security-relevant behavior.

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

File: src/lib/ai/client.ts [hunk 1]
```diff
@@ -19,6 +19,43 @@ export class MissingApiKeyError extends Error {
...
+  }
+  if (
+    !(unwrapped instanceof Error) &&
+    typeof unwrapped?.message === "string" &&
+    unwrapped.message.trim()
+  ) {
+    return unwrapped.message;
+  }
+  return errorMessage(e);
+}
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/lib/ai/client.ts:69]: Accepted — with a correction to the mechanism, and it pointed at a better fix than the one implied.

Two parts of the diagnosis don't hold against the code as reviewed:

- **The object path did exist.** Lines 49-55 read `!(unwrapped instanceof Error) && typeof unwrapped?.message === "string"`, which returns the message for the SDK's actual in-stream shape. `@ai-sdk/openai` enqueues `{type:"error", error: value.error}` — the already-unwrapped `{message, type, param, code}` — so `[object Object]` wasn't reachable on that path.
- **`responseBody` is always a string here.** It's built by `readResponseBodyAsText()` (`@ai-sdk/provider-utils/dist/index.mjs:2729`), so the non-string-body case doesn't arise in this SDK.

But the underlying point was right: the shape handling didn't match what the comment promised. Executing all four combinations showed a real gap:

| payload | before |
| --- | --- |
| string body `{error:{message}}` | ✅ |
| string body bare `{message}` | ❌ "Bad Request" |
| object payload `{message}` | ✅ |
| object payload `{error:{message}}` | ❌ "Bad Request" |

Each path handled exactly one shape. A bare `{"message":…}` body — plausible from an OpenAI-compatible server — lost the reason, which is precisely the bug this PR exists to fix.

Fixed in the next push: both paths now share one reader (`errorTextOf`) that accepts `{error:{message}}` or a bare `{message}`, array-wrapped or not. All four combinations pass, alongside Google's array-wrapped 400, the `RetryError`→`lastError` 429, and the fallback cases — 11 shapes total.

The doc comment has also been rewritten to state the constraints rather than describe the paths, so it can't drift from the behaviour again.


---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
## Round 1 dispositions

Copilot's inline finding is answered and resolved in-thread (accepted with a mechanism correction — it found a real shape asymmetry, though not the one it described). The security audit returned clean. Below is the AI review, item by item.

### should-fix — allowlist writers don't discard the verdict → **FIXED** (next push)

Confirmed by reading, not taken on trust: `testConnection` passes the **draft** `allowedHosts` into `createAiClient` (`AiProviderSection.tsx:728-732`), and both writers bypassed `discardTestResult` — `allowHost` (`:700-705`) and the `AllowedHostsField` `onChange` (`:1064`). Both now discard, with the reason recorded at the call site.

Worth saying plainly: **this is the same class as the bug the PR sets out to fix, and my own pre-open class sweep missed it.** I answered the "what dimension is this rule silent about?" question with *non-HTTP provider failures*, when the live silent dimension was *other inputs to the verdict*. Good catch.

So I re-swept by **input** rather than by symbol, and enumerated every writer of anything `testConnection` consumes:

| input | writers | status |
| --- | --- | --- |
| `ai` draft (provider, model, base URLs) | exactly one — `setAi` (`:693`); all three generation surfaces route through it | already discards |
| `aiAllowedHosts` | `allowHost` (`:703`), `AllowedHostsField onChange` (`:1064`) | both now discard |
| typed key | `keyForm` | declined, below |
| `reviewAi` / `securityReviewAi` (`:911`–`:983`) | separate fields | not inputs — `testConnection` only tests the generation provider |

### Deliberate call — `McpServersSection.tsx:127`

You're right that it writes the same list with no access to this state. Recording it rather than leaving it silent: closing it means lifting the test-result state out of `AiProviderSection`, which is a structural change well beyond this PR. The practical exposure is narrow — reaching it requires editing MCP allowed-hosts in one section while a stale AI connection verdict sits in another.

### nit — fallback returns the wrapper → **ACCEPTED as a conscious choice** (next push)

Kept as `errorMessage(e)` deliberately, now stated in-code: the retry message embeds the last error's text *and* the attempt count (`@ai-sdk/provider-utils/dist/index.mjs:2672`), and the count is worth keeping. Previously accidental, now recorded.

### nit — doc comment argues the diff → **ACCEPTED** (next push)

Rewritten as three constraints, and your second point was the sharper one: the `message` read is a fallback, not an unwrap, and calling it one was simply wrong.

### nit — clear on key-field typing → **DECLINED, with evidence**

Two reviewers raised this independently, so I checked the mechanism rather than the idea — and the suggested mechanism isn't verified in this codebase. `listeners` has **zero** uses anywhere in `src/`, and `TextField`'s props are closed with no change hook (`components/form/fields.tsx:35-55`), so `listeners={{ onChange: … }}` would be an unproven API on a shared form primitive.

The scope line: the verdict clears when the **configuration it described** changes — provider, model, allow list, saved key. A typed-but-unsaved key hasn't changed the configuration yet; it becomes real at Save, which already discards.

**Reversal shape** — if you'd rather have it, the verified execution is a `useSelector` subscription on `keyForm.store` (an idiom this file already uses for every other draft value) or an explicit `onChange` prop added to `TextField`. Not a side effect inside `validators`, which is where it would most naturally be reached for.

---

Gates on the batch: `tsc -b --noEmit` exit 0; biome format (EOL-pinned) + check clean; `changelog:check` 4 valid; the error reader re-verified by execution over 11 payload shapes.


---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

Two follow-up fixes landed cleanly (both allowlist writers now discard, and the fallback choice is recorded in-code), and the shared `errorTextOf` reader is a real improvement over the two single-shape paths. Two things worth addressing before merge-quality: the rewritten helper still misses Ollama's native error body shape, and the rewrite left the `providerErrorMessage` doc block attached to the wrong symbol.

### Correctness / edge cases

- **should-fix** — `src/lib/ai/client.ts:35-42`, `errorTextOf`: it reads `entry?.error?.message ?? entry?.message`, so a body whose `error` is a bare **string** yields `null` and the caller falls through to the generic message. That's Ollama's native shape, and Ollama is a first-class provider here: `model-factory.ts:51-58` points both `ollama` and `ollama-cloud` at `${base}/api` (the native endpoint, not `/v1`), and `ollama-ai-provider-v2/dist/index.js:221-234` hard-requires `error` to be an *object* (`z.object({ error: z.object({ message: z.string(), … }) })`). Concrete case: Test connection against a model the user hasn't pulled → Ollama answers 404 with `{"error":"model \"llama3\" not found, try pulling it first"}` → the schema parse fails → `createJsonErrorResponseHandler` sets `APICallError.message = response.statusText` and keeps the raw body in `responseBody` (`@ai-sdk/provider-utils@4.0.38/dist/index.mjs:2745-2775`) → `errorTextOf` returns `null` → the banner reads "Not Found". That is precisely the failure class this PR exists to close, on the provider most likely to hit it. Fix inside the same helper:
  ```ts
  const entry = (Array.isArray(value) ? value[0] : value) as {
    error?: { message?: unknown } | string;
    message?: unknown;
  } | null;
  const nested = entry?.error;
  const message =
    typeof nested === "string" ? nested : (nested?.message ?? entry?.message);
  ```
  Extend `errorTextOf`'s own doc comment's shape list in the same edit (it currently enumerates "nests under `error` … or sits bare on `message`") to include the bare string under `error` (Ollama's native `/api`), so the comment still matches what the reader accepts. No other call site or import changes.

### Readability

- **should-fix** — `src/lib/ai/client.ts:22-35`: the rewritten `providerErrorMessage` block comment (lines 22-30) is now immediately followed by a second `/** … */` block (lines 31-34) and then `function errorTextOf`. JSDoc binds to the nearest preceding block, so `errorTextOf` owns lines 31-34 and lines 22-30 are orphaned — hovering `providerErrorMessage` (line 44) shows no documentation at all, and the "three constraints" text now sits above a helper it doesn't describe. This is collateral from the last push (the helper was inserted between the doc block and its function). Fix: move the lines 22-30 block down so it sits directly above `function providerErrorMessage`, leaving `errorTextOf`'s block as the only comment above `errorTextOf`.

- **nit** — `changelog.d/fixed-ai-connection-test-feedback.md`: the bullet scopes the improvement to **Test connection**, but `providerErrorMessage` is wired into the generation stream (`client.ts:149`, `:172`) and the agentic review path (`:288`, `:351`, `:365`) too, so a failed commit-message generation or AI review now surfaces the provider's reason as well. Worth widening the first clause (e.g. "AI failures explain themselves — a failed **Test connection**, generation or review now shows the provider's own reason…") rather than shipping a fragment narrower than the change. The recorded "no README / help / site surface changed" call still checks out — nothing in those enumerates model ids or "Test connection" copy.

Recorded decisions I'm not re-raising: clearing the verdict on typed-but-unsaved key edits (declined with a verified mechanism note), and `McpServersSection.tsx:127` writing the same allowlist without access to this state (deliberate, structural to fix).

### Resolved since last review

- Both allowlist writers now discard the verdict — `allowHost` (`AiProviderSection.tsx:702-707`) and `AllowedHostsField onChange` (`:1063-1066`), each with the reason at the call site.
- The wrapper-vs-unwrapped fallback at `client.ts:63-65` is now a stated choice (attempt count kept) rather than an accident.
- The doc comment no longer argues the diff; it states constraints (placement issue above notwithstanding).
- Copilot's `client.ts:66` shape finding: the object-payload path and the string-body path now share one reader, so `{error:{message}}` and bare `{message}` are accepted on both — no longer applicable.

Verdict: no blocking issues — remaining items are non-blocking; merge when ready

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
## Round 2 dispositions

Both should-fixes accepted — and both were independently confirmed by a concurrent second read before this reply, the first one with a live probe.

### should-fix — Ollama's string-`error` shape → **FIXED** (next push)

Not just inferred from the zod schema: measured against a running Ollama daemon —

```
POST /api/chat  {"model":"nonexistent-model-xyz",…}
→ {"error":"model 'nonexistent-model-xyz' not found"}
```

Bare string under `error`, exactly as you said, and `ollama-ai-provider-v2`'s schema does hard-require an object — so the parse fails and the banner would read "Not Found" on the provider most likely to hit this. Your suggested reader was re-derived against the file and adopted as-is, with `errorTextOf`'s shape-list comment extended in the same edit. The verification harness now covers 14 shapes, including this measured body, `{error:"E"}` as a raw payload, and the blank-string / non-string-message fallbacks.

### should-fix — stranded doc block → **FIXED** (next push)

Correct on the mechanism (JSDoc binds to the nearest preceding block) and correct that it was collateral from the last push — the helper was inserted between the comment and its function. Reordered: `errorTextOf` + its comment first, `providerErrorMessage`'s block moved down to sit on its function. Adjacency re-verified by grep after the move.

### nit — fragment narrower than the change → **ACCEPTED, widened on both ends**

You flagged the "Test connection"-only clause; the second read independently flagged the *other* clause — "clears when you save or remove a key" was also narrower than the shipped clearing set (provider/model switches and allow-list edits clear too). One rewrite covers both: "A failed **Test connection**, generation, or AI review…" and "clears whenever what it tested changes — key saved or removed, provider or model switched, allowed hosts edited."

---

Gates on the batch: `tsc -b --noEmit` exit 0 · biome (EOL-pinned) clean · `changelog:check` 4 valid · 14/14 harness. Standing declines unchanged and not re-argued. Lands on the next push.


---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

Both prior should-fixes are fixed cleanly and the fix hunks themselves are correct — `errorTextOf` now narrows `error` as `string | {message?}` in a way TypeScript accepts, and the reorder left `client.ts`'s imports and the `errorMessage` fallback (`:68`) intact. One new should-fix: a stale verdict still survives the Settings screen's **Discard**, which is the same class this PR exists to close.

### Correctness / edge cases

- **should-fix** — `src/features/settings/AiProviderSection.tsx:665-668` (`discardTestResult`) and its call sites `:694`, `:706`, `:1065`: the verdict is cleared imperatively at each *in-section* writer, so any write to the same draft from outside the section leaves it pinned. Concrete path, all in code shown or read: on the AI panel switch the generation model (→ `setAi` discards, `dirty` flips, the footer appears), click **Test connection** → green "Connected" renders at `:874-878`, then click the footer's **Discard** (`SettingsScreen.tsx:316-318` → `discard()` at `:202-204` → `form.reset(saved, { keepDefaultValues: true })`). That reverts `values.ai`/`values.aiAllowedHosts` while `AiProviderSection` stays mounted (`SettingsScreen.tsx:261-266` renders it plainly under `activePanel === "ai"`, no key, no unmount), so "Connected" now describes the model you just discarded — and the fragment's claim "clears whenever what it tested changes — … provider or model switched" is false on that path. (The dialog's *Discard and close* path is fine: it calls `closeSettings()`, which unmounts the section.)

  Fix: derive validity instead of clearing for draft inputs — store the configuration under test with the result and render only while it still matches.
  ```tsx
  const [testResult, setTestResult] = useState<{
    ok: boolean; message?: string; config: string;
  } | null>(null);
  // Signature over every draft input testConnection reads. A key-order difference can
  // only clear a verdict early, never keep a stale one.
  const testConfig = JSON.stringify({ ai, allowedHosts });
  ```
  In `testConnection` capture `const config = testConfig;` before the first `await` and pass it into both `setTestResult` calls (`:738`, `:743`); gate both render branches (`:874`, `:879`) on `testResult.config === testConfig`. Knock-ons to apply in the same edit: delete the now-redundant `discardTestResult()` calls in `setAi` (`:694`), `allowHost` (`:706`) and `AllowedHostsField onChange` (`:1065`) together with their explanatory comments (which would otherwise describe code that no longer clears anything); keep `testRun` plus the run-guards at `:737`/`:742` and the two key-path discards (`keyForm.onSubmit` `:680`, `clearKey` `:717`), since the keychain key isn't part of the draft signature; don't reach for `stableStringify` — it's module-local to `SettingsScreen.tsx:80` and unexported. This also closes the `McpServersSection.tsx:127` allow-list writer without lifting state out of the section, so that standing recorded decision (deliberate, "structural to fix") could be retired rather than carried forward.

Documentation: the two fragments cover the user-facing half, and the recorded "no README / site / help change" call still holds — I grepped all four surfaces for Gemini ids and none enumerates a model (`README.md:817`, `site/src/data/capabilities.ts:641`, `site/src/pages/ai.astro:14`, `src/features/help/content.ts:1782-1784` all speak of the provider, not ids).

### Resolved since last review

- Ollama's bare-string `error` body is now read (`client.ts:27-36`), and `errorTextOf`'s own shape-list comment was extended in the same edit to name it (`:22-26`) — no drift between reader and comment.
- The stranded doc block is fixed: `errorTextOf` carries its own comment, and the "three constraints" block (`:38-46`) now sits directly above `function providerErrorMessage` (`:47`). The reorder introduced no collateral — imports unchanged, `errorMessage` still imported at `:9` and still the fallback at `:68`.
- `changelog.d/fixed-ai-connection-test-feedback.md` now covers generation and AI review as well as **Test connection**, and states the full clearing set.

Verdict: no blocking issues — remaining items are non-blocking; merge when ready

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
## Round 3 disposition

### should-fix — verdict survives the footer Discard → **FIXED as prescribed** (next push)

Every mechanism claim verified before adopting: `discard()` is `form.reset(saved)` at `SettingsScreen.tsx:202-204` with the section still mounted (`:261-266`, no key), both render branches were ungated, and `stableStringify` is indeed module-local.

Accepted the derived-validity design wholesale rather than adding a fourth clear site — the imperative approach was the bug factory here: round 1 added two clear calls, this round found a third route, and `McpServersSection` was a known fourth. The verdict now carries the draft signature it was produced under (captured before the first await) and renders only while the live signature matches.

Knock-ons applied exactly as enumerated: the three draft-writer discards are deleted with their comments; `testRun` and the run-guards stay (the keychain key is invisible to the draft signature, so save/clear keep their imperative discards); plain `JSON.stringify` (same-literal construction gives stable key order, and a mismatch can only hide a verdict early, never keep a stale one).

**Standing decision retired:** the `McpServersSection.tsx:127` allow-list writer — previously recorded as "deliberate, structural to fix" — is now covered for free, since its write changes the signature.

One behavior noted for the record, not a defect: switching a model away and back resurrects the verdict. That's the semantics of the design — the verdict is a memo of "this exact configuration tested green" — and the fragment's "clears whenever what it tested changes" stays true user-visibly.

---

Gates: `tsc -b --noEmit` exit 0 · biome (EOL-pinned) clean · `changelog:check` 4 valid. What static review can't see on this one: the rendered behavior of the Discard path itself — verified by code walk, not by driving the settings screen live.


---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

The round-3 fix lands the derived-validity design as prescribed: the verdict now carries the draft signature it was produced under and renders only while it still matches, so the footer **Discard** path no longer pins a stale "Connected". I re-read the fix hunks as new code and found no collateral — no blockers.

### Resolved since last review

- Stale verdict surviving the footer **Discard** — fixed at `AiProviderSection.tsx:668` (`testConfig`), `:733` (captured before the first `await`), `:746-753` (both `setTestResult` calls carry `config`) and `:884`/`:889` (both render branches gated). `form.reset(saved)` at `SettingsScreen.tsx:203` changes `values.ai`/`values.aiAllowedHosts`, the section's `useSelector`s (`:598`, `:627`) re-derive `testConfig`, and the verdict retires without the section knowing to clear it.
- Fix hunks reviewed on their own terms, no collateral: the three draft-writer `discardTestResult()` calls are gone with their comments (`setAi` `:702-704`, `allowHost` `:708-713`, `AllowedHostsField onChange` `:1073`); `discardTestResult` still has two live callers (`:689`, `:722`), `useRef` was already imported (`:12`), and `errorMessage` is still imported (`:67`) and still used at `:753`. The rewritten `testRun` comment (`:669-671`) matches those two remaining callers — keys really are invisible to the signature, and `setSecret`/`deleteSecret` have no other call sites in `src/` (`git/api.ts:3370,3380` are the only definitions; `RepoSettingsDialog.tsx:798` is an unrelated local setter), so no external writer can leave a verdict describing a superseded key.
- The `McpServersSection.tsx:127` standing decision is genuinely retired, not just claimed: it writes the same `aiAllowedHosts` field on the shared form, so the signature moves — and the panels are mutually exclusive (`SettingsScreen.tsx:261-270`), which unmounts the section anyway.
- Earlier rounds' items remain fixed in the current file: Ollama's bare-string `error` (`client.ts:27-36`), the `errorTextOf`/`providerErrorMessage` comment adjacency (`:22-26`, `:38-47`), and the widened `changelog.d/fixed-ai-connection-test-feedback.md`.

The "switch away and back resurrects the verdict" trade-off is a recorded decision and holds up under the code: any key change clears `testResult` outright, so a resurrected verdict can only describe a configuration that tested green against the current key. Documentation: the two fragments cover the user-facing half; the recorded "no README / site / in-app guide change" call still stands per last round's grep of those surfaces.

### Leftover polish (non-blocking)

- `src/lib/ai/client.ts:68` (unchanged since the last review) — a non-`Error` payload whose shape `errorTextOf` doesn't recognize (no `message`, no `error.message`, no string `error`) falls to `errorMessage(e)` → `String(e)` → `[object Object]`; a `JSON.stringify` fallback for non-`Error` objects would keep the raw payload readable, in the spirit of the rest of the helper.

Nothing further to raise.

Verdict: no blocking issues — remaining items are non-blocking; merge when ready

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
## Round 4 disposition — converged

### nit — unrecognizable payload renders `[object Object]` → **deferred, with evidence and a home**

Checked the trigger set rather than the plausibility: OpenAI and Anthropic schema-validate their in-stream error chunks as message-bearing objects before enqueueing, and Ollama's raw-value fallback enqueues its actual response body — `{"error":"string"}`, which `errorTextOf` reads as of round 2. So the shape this guards against — no `message`, no `error.message`, no string `error` — is emitted by nothing installed. A `JSON.stringify` fallback would trade a today-impossible `[object Object]` for raw JSON in user-facing banners on any future odd shape.

Home: it rides the backlogged model-catalog error-surface work, which touches the same helper. If it's wanted sooner, the right execution is a capped or dev-only stringify — never raw JSON in the banner.

### Convergence

This round found nothing real, and it closes the loop:

- Round 4 re-verified every prior fix in the current file, including the sharper proofs — no external `setSecret`/`deleteSecret` writer exists, and the resurrect-on-revert trade-off is safe because key changes clear outright.
- CI is green on this head (`build`, `fragment`, Pages).
- One full clean round after the last substantive fix.
- Copilot's single thread: resolved in round 1. Security audit: clean. Every review item since: dispositioned on the record.

**Merge-ready.** Four rounds, four pushes, each push carrying a real fix — nothing outstanding beyond the two deferred items already living on the backlog (model-catalog error surface, and this stringify nit folded into it).


---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._
